raulin0/backup-juggler
# feat: adds backup feature

## Description
This pull request includes a series of changes and additions aimed at improving the project configuration, documentation, testing, and CI/CD pipeline. The updates provide enhanced functionality, improved documentation, and automated testing processes.

## Changes Made
- Updated the project configuration, metadata, and documentation URLs to ensure accurate and up-to-date information.
- Added a .readthedocs.yml configuration file to facilitate the generation of documentation using Read the Docs (RTD).
- Updated various documentation files, including the contribution guide, index page, tutorial, installation instructions, and API documentation for Backup Juggler.
- Updated the MkDocs configuration file to reflect changes in the documentation generator tool.
- Implemented additional tests for the Backup Juggler CLI and added a fixture for running CLI tests.
- Added the Backup Juggler CLI functionality to allow users to interact with the project through the command line.
- Implemented unit tests for the backup functionality and added fixtures for temporary paths to ensure reliable testing.
- Implemented the file backup functionality with a progress viewer and introduced a paths model class for managing file paths and directories during the backup process.
- Set up a GitHub Actions pipeline for testing and code coverage, ensuring the project's functionality remains intact and maintaining a certain level of code coverage.

## Motivation and Context
These changes were made to improve the project's overall functionality, maintainability, and documentation. The Backup Juggler project benefits from enhanced configuration, automated testing, improved documentation, and the addition of a CLI for user convenience. By making these improvements, the project becomes more user-friendly, easier to contribute to, and better equipped for ongoing maintenance.

## How to Test
1. Review the updated project configuration, ensuring that metadata and documentation URLs are accurate.
2. Verify the presence and correctness of the .readthedocs.yml configuration file for Read the Docs.
3. Read through the updated documentation files, including the contribution guide, index, tutorial, installation instructions, and API documentation for Backup Juggler.
4. Check that the MkDocs configuration has been properly updated to reflect any changes made.
5. Run the additional tests for the Backup Juggler CLI functionality and confirm they pass successfully.
6. Test the Backup Juggler CLI by interacting with it through the command line, ensuring that the expected functionality is available and functioning correctly.
7. Execute the unit tests for the backup functionality and verify that they pass, using the added fixtures for temporary paths.
8. Perform a file backup operation using the updated functionality and confirm that the progress viewer accurately reflects the progress of the backup process.
9. Check that the paths model class correctly manages file paths and directories during the backup operation.
10. Validate the GitHub Actions pipeline by ensuring that tests are executed and code coverage reports are generated successfully.

## Related Issue(s)
N/A

## Checklist

Please review and merge this pull request at your earliest convenience. Feel free to provide any feedback or suggestions for improvement. Thank you!
